src/Workspaces/CoreTestUtilities/MEF/TestComposition.cs

```diff
@@ -0,0 +1,223 @@
+
+            public override int GetHashCode()
+                => Hash.Combine(Hash.Combine(Hash.CombineValues(_assemblies), Hash.CombineValues(_parts)), Hash.CombineValues(_excludedPartTypes));
+
+            public static bool operator ==(CacheKey left, CacheKey right)
+                => left.Equals(right);
+
+            public static bool operator !=(CacheKey left, CacheKey right)
+                => !(left == right);
+        }
+
+        /// <summary>
+        /// Assemblies to include in the composition.
+        /// </summary>
+        public readonly ImmutableHashSet<Assembly> Assemblies;
+
+        /// <summary>
+        /// Types to exclude from the composition.
+        /// All subtypes of types specified in <see cref="ExcludedPartTypes"/> and defined in <see cref="Assemblies"/> are excluded before <see cref="Parts"/> are added.
+        /// </summary>
+        public readonly ImmutableHashSet<Type> ExcludedPartTypes;
+
+        /// <summary>
+        /// Additional part types to add to the composition.
+        /// </summary>
+        public readonly ImmutableHashSet<Type> Parts;
+
+        private readonly Lazy<IExportProviderFactory> _exportProviderFactory;
+
+        private TestComposition(ImmutableHashSet<Assembly> assemblies, ImmutableHashSet<Type> parts, ImmutableHashSet<Type> excludedPartTypes)
+        {
+            Assemblies = assemblies;
+            Parts = parts;
+            ExcludedPartTypes = excludedPartTypes;
+
+            _exportProviderFactory = new Lazy<IExportProviderFactory>(GetOrCreateFactory);
+        }
+
+        /// <summary>
+        /// Returns a new instance of <see cref="HostServices"/> for the composition. This will either be a MEF composition or VS MEF composition host, 
+        /// depending on what layer the composition is for. Editor Features and VS layers use VS MEF composition while anything else uses System.Composition.
+        /// </summary>
+        public HostServices GetHostServices()
+            => VisualStudioMefHostServices.Create(ExportProviderFactory.CreateExportProvider());
+
+        /// <summary>
+        /// VS MEF <see cref="ExportProvider"/>.
+        /// </summary>
+        public IExportProviderFactory ExportProviderFactory => _exportProviderFactory.Value;
+
+        private IExportProviderFactory GetOrCreateFactory()
+        {
+            var key = new CacheKey(Assemblies, Parts, ExcludedPartTypes);
+
+            lock (s_factoryCache)
+            {
+                if (s_factoryCache.TryGetValue(key, out var existing))
+                {
+                    return existing;
+                }
+            }
+
+            var newFactory = ExportProviderCache.CreateExportProviderFactory(GetCatalog(), IsRemote);
+
+            lock (s_factoryCache)
+            {
+                if (s_factoryCache.TryGetValue(key, out var existing))
+                {
+                    return existing;
+                }
+
+                s_factoryCache.Add(key, newFactory);
```

Comment:
💡 I believe we have a `GetOrAdd` extension method for this

Reply:
We have `GetOrAdd` that takes a lambda.